cpp/src_prims/random/make_regression.h

```diff
@@ -0,0 +1,290 @@
+#include <cuml/common/cuml_allocator.hpp>
+
+#include "linalg/add.h"
+#include "linalg/cublas_wrappers.h"
+#include "linalg/init.h"
+#include "linalg/qr.h"
+#include "linalg/transpose.h"
+#include "matrix/matrix.h"
+#include "permute.h"
+#include "rng.h"
+#include "utils.h"
+
+namespace MLCommon {
+namespace Random {
+
+/* Internal auxiliary function to help build the singular profile */
+template <bool square, typename DataT, typename IdxT>
+static __global__ void _singular_profile_kernel(DataT* out, IdxT n,
+                                                DataT strength, DataT coef,
+                                                IdxT rank) {
+  IdxT tid = threadIdx.x + blockIdx.x * blockDim.x;
+  if (tid < n) {
+    DataT sval = static_cast<DataT>(tid) / rank;
+    if (square) sval *= sval;
+    out[tid] = strength * myExp(coef * sval);
+  }
+}
+
+/* Internal auxiliary function to generate a low-rank matrix */
+template <typename DataT, typename IdxT>
+static void _make_low_rank_matrix(DataT* out, IdxT n_rows, IdxT n_cols,
+                                  IdxT effective_rank, DataT tail_strength,
+                                  Rng& r, cublasHandle_t cublas_handle,
+                                  cusolverDnHandle_t cusolver_handle,
+                                  std::shared_ptr<deviceAllocator> allocator,
+                                  cudaStream_t stream) {
+  IdxT n = std::min(n_rows, n_cols);
+
+  // Generate random (ortho normal) vectors with QR decomposition
+  device_buffer<DataT> rd_mat_0(allocator, stream);
+  device_buffer<DataT> rd_mat_1(allocator, stream);
+  rd_mat_0.resize(n_rows * n, stream);
+  rd_mat_1.resize(n_cols * n, stream);
+  r.normal(rd_mat_0.data(), n_rows * n, (DataT)0.0, (DataT)1.0, stream);
+  r.normal(rd_mat_1.data(), n_cols * n, (DataT)0.0, (DataT)1.0, stream);
+  device_buffer<DataT> q0(allocator, stream);
+  device_buffer<DataT> q1(allocator, stream);
+  q0.resize(n_rows * n, stream);
+  q1.resize(n_cols * n, stream);
+  LinAlg::qrGetQ(rd_mat_0.data(), q0.data(), n_rows, n, cusolver_handle, stream,
+                 allocator);
+  LinAlg::qrGetQ(rd_mat_1.data(), q1.data(), n_cols, n, cusolver_handle, stream,
+                 allocator);
+
+  // Build the singular profile by assembling signal and noise components
+  device_buffer<DataT> low_rank(allocator, stream);
+  device_buffer<DataT> tail(allocator, stream);
+  device_buffer<DataT> singular(allocator, stream);
+  low_rank.resize(n, stream);
+  tail.resize(n, stream);
+  _singular_profile_kernel<true><<<ceildiv<IdxT>(n, 256), 256, 0, stream>>>(
+    low_rank.data(), n, (DataT)1.0 - tail_strength, (DataT)-1.0,
+    effective_rank);
+  CUDA_CHECK(cudaPeekAtLastError());
+  _singular_profile_kernel<false><<<ceildiv<IdxT>(n, 256), 256, 0, stream>>>(
+    tail.data(), n, tail_strength, (DataT)-0.1, effective_rank);
+  CUDA_CHECK(cudaPeekAtLastError());
+  LinAlg::add(low_rank.data(), low_rank.data(), tail.data(), n, stream);
```

Comment:
the `_singular_profile_kernel` can be updated to compute `low_rank`, `tail` and their summation in a single call, no? It reduces that many global memory traffic and also the need to create another array.

Reply:
Done. I haven't really looked at the performance yet, I can do some profiling and optimization if it's important.